### PR TITLE
Update HttpClientJsonExtensions.Get.AsyncEnumerable.cs

### LINKED ISSUE #92179: `GetFromJsonAsAsyncEnumerable` extension default json serialize option is not the same as `GetFromJson`
### Description

`GetFromJsonAsAsyncEnumerable` extension default json serialize option is not the same as `GetFromJson`, for `GetFromJsonAsync<TValue>` etc extensions using the `new JsonSerializerOptions(JsonSerializerDefaults.Web)`, while the new `GetFromJsonAsAsyncEnumerable` extensions using the `JsonSerializerOptions.Default`

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.cs#L19

https://github.com/dotnet/runtime/blob/fa1bf3c0f9ec4e204203a49191170f9d53cd549a/src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonHelpers.cs#L15C10-L15C10

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs#L138

https://github.com/dotnet/runtime/pull/89258/files#diff-096816e722c9256aa978e030242edf422dd8e41be616ab27bd7175cd522789e2R138

### Reproduction Steps

```c#

using var httpClient = new HttpClient();
httpClient.BaseAddress = new Uri("http://localhost:5297");
// we assume the response returned like
// [{"id":1,"title":"job_0 --- 09/16/2023 21:32:26 +08:00"},{"id":2,"title":"job_1 --- 09/16/2023 21:32:27 +08:00"}]
var stream = httpClient.GetFromJsonAsAsyncEnumerable<Job>("api/Jobs");
await foreach (var job in stream)
{
    Console.WriteLine(job);
    Console.WriteLine(DateTimeOffset.Now);
}

public sealed class Job
{
    public required int Id { get; set; }
    public required string Title { get; set; }
    public override string ToString() => $"JobId: {Id}, JobTitle: {Title}";
}
```

### Expected behavior

Think it should behave like `GetFromJsonAsync` to use the `JsonSerializerDefault.Web`

### Actual behavior

image

### Regression?

_No response_

### Known Workarounds

The `HttpContent` extension `ReadFromJsonAsAsyncEnumerable` is working

```c#
using var httpClient = new HttpClient();
httpClient.BaseAddress = new Uri("http://localhost:5297");
using var response = await httpClient.GetAsync("api/Jobs", HttpCompletionOption.ResponseHeadersRead);
ArgumentNullException.ThrowIfNull(response.Content);
var stream = response.Content.ReadFromJsonAsAsyncEnumerable<Job>();
await foreach (var job in stream)
{
    Console.WriteLine(job);
    Console.WriteLine(DateTimeOffset.Now);
}
```

### Configuration

.NET 8 RC 1

### Other information

_No response_

### PR DESCRIPTION
**Pull request by WeihanLi**
Update `jsonTypeInfo` for `ReadFromJsonAsAsyncEnumerable`

fixes #92179

### CODE REVIEW

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs [hunk 1]
```diff
@@ -166,11 +183,8 @@ public static partial class HttpClientJsonExtensions
                     .ConfigureAwait(false);
                 response.EnsureSuccessStatusCode();
 
-                using Stream readStream = await GetHttpResponseStreamAsync(client, response, false, cancellationToken)
-                    .ConfigureAwait(false);
-
-                await foreach (TValue? value in JsonSerializer.DeserializeAsyncEnumerable<TValue>(
-                    readStream, jsonTypeInfo, cancellationToken).ConfigureAwait(false))
+                await foreach (TValue? value in response.Content.ReadFromJsonAsAsyncEnumerable<TValue>(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs:186]: Please revert this change, it is unrelated to the fix at hand and the `GetHttpResponseStreamAsync` uses logic that is subtly different to reading directly from `HttpContent` values.

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs [hunk 2]
```diff
@@ -135,12 +135,29 @@ public static partial class HttpClientJsonExtensions
...
+            if (client is null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
 
-            var jsonTypeInfo = (JsonTypeInfo<TValue>)options.GetTypeInfo(typeof(TValue));
+            return Core(client, requestUri, options, cancellationToken);
 
-            return FromJsonStreamAsyncCore(client, requestUri, jsonTypeInfo, cancellationToken);
+            static async IAsyncEnumerable<TValue?> Core(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs:145]: Likewise, this change would need to be reverted.

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs [hunk 3]
```diff
@@ -135,12 +135,29 @@ public static partial class HttpClientJsonExtensions
             JsonSerializerOptions? options,
             CancellationToken cancellationToken)
         {
-            options ??= JsonSerializerOptions.Default;
-            options.MakeReadOnly();
+            if (client is null)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs:138]: Private helpers should not be doing null validation. I would recommend replacing this will a `Debug.Assert(client != null);` statement and make the actual checks in the public methods that call into the helpers. This also avoids redundant checking that `client` is null in cases where the IAsyncEnumerable will be iterated over multiple times.

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs [hunk 4]
```diff
@@ -135,12 +135,29 @@ public static partial class HttpClientJsonExtensions
...
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
 
-            var jsonTypeInfo = (JsonTypeInfo<TValue>)options.GetTypeInfo(typeof(TValue));
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs:141]: The crux of the fix should be replacing the above with a call to the `JsonHelpers.GetJsonTypeInfo` method. See the `HttpContent` implementation for pointers.

File: src/libraries/System.Net.Http.Json/tests/FunctionalTests/HttpClientJsonExtensionsTests.cs [hunk 5]
```diff
@@ -580,6 +580,25 @@ await foreach (string s in client.GetFromJsonAsAsyncEnumerable<string>("http://d
                 await Task.Delay(TimeSpan.FromMilliseconds(10));
             }
         }
+
+        [Fact]
+        public async Task GetFromJsonAsAsyncEnumerable_SerializerCaseInsensitiveTest()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Http.Json/tests/FunctionalTests/HttpClientJsonExtensionsTests.cs:585]: Technically web defaults use camel casing, not case insensitive binding.
```suggestion
        public async Task GetFromJsonAsAsyncEnumerable_SerializerUsesCamelCase()
```

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs (hunk 1 above)
Comment (WeihanLi) [reply src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs]: Reverted

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs (hunk 2 above)
Comment (WeihanLi) [reply src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs]: Reverted

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs (hunk 3 above)
Comment (WeihanLi) [reply src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs]: updated

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs (hunk 4 above)
Comment (WeihanLi) [reply src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.Get.AsyncEnumerable.cs]: updated

**Comment by WeihanLi** [comment]
Seemed the failed CI not related
image

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!
